### PR TITLE
Create a separate bundle for cart and checkout

### LINKED ISSUE #48305: Create a separate Webpack bundle for the cart and checkout frontend blocks
**Issue Description:**

It was found that if we separate out the cart and checkout frontend scripts to their own Webpack bundle it will allow a reduction in bundle size. The reduction in bundle size will come from more efficient vendor bundles being split out between all the non cart/checkout blocks and the cart/checkout.

Because cart/checkout is basically a React app composed of blocks, it won't often be loaded alongside regular blocks, the vendor dependencies it needs are different from those needed by the general blocks.

Also it will allow the cart/checkout to remove old dependencies to further shave size off the bundle without worrying about where those dependencies are used in existing non cart/checkout blocks.

### PR DESCRIPTION
**Pull request by samueljseay**
### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/48305

This PR separates out a bundle for the cart and checkout blocks with the following changes:

1. Create a new Webpack bundle, adjusting the Webpack entry point logic to separate cart and checkout entry points from the regular block ones
2. Create a new Webpack plugin that we run on regular frontend bundle and the new one. The plugin adds vendor dependency handles to the asset.php files generated. This ensures we don't need specific enqueue logic for assets in AbstractBlock or anywhere else. Entrypoints will always load the vendors generated from their bundles. 

The result: 

* Trim 112KB across the old frontend bundle and the new cart and checkout bundle
* Cart page total JS loaded decreases ~30KB
* Checkout page total JS loaded decreases ~40KB

### How to test the changes in this Pull Request:

* CI should pass for all the blocks Playwright tests

#### Smoke test

* A smoke test of functionality in cart and checkout, especially check that mini cart still functions by: open mini cart, close mini cart, add, remove products etc

#### Jetpack Boost 3.3.1

Jetpack Boost 3.3.1 is a slightly older version which does not exclude minifying these new scripts. In this PR we add an exclusion via a Jetpack Boost filter (when the older version is present) to ensure it doesn't break our scripts by re-minifying them.

To test with Jetpack Boost 3.3.1: 

1. Download the older version of JP boost: https://downloads.wordpress.org/plugin/jetpack-boost.3.3.1.zip 
2. Run a JN site using this branch. Install the older JP boost plugin
3. From admin go to Jetpack > Boost  and enable "Concatenate JS" toggle

Screenshot 2024-06-12 at 10 32 50 AM

3. Smoke test cart and checkout once more, we're looking for critical errors or white screens.

4. (optional) Also you can see that the code has an effect by disabling it. Go to plugin file editor and edit WooCommerce > src > Blocks > AssetController > `skip_boost_minification_for_cart_checkout`. Change it to return `$do_concat`. Then try load cart and checkout and note they fail to load and there is a console error.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Create a separate JS cart and checkout JavaScript bundle to improve performance.

</details>

<details>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Blocks/BlockTypes/AbstractBlock.php
Comment (samueljseay) [inline plugins/woocommerce/src/Blocks/BlockTypes/AbstractBlock.php]: This is a bit sus, adding vendor deps can be quite manual. I'd love to build some concept of the bundle in PHP rather than what we have in AssetsController right now but it'd be quite a refactor.

We need to ensure that only non cart and checkout blocks load this vendor script, we could use class inheritance instead somehow? ie we could include this by default in AbstractBlock but have  `Cart.php` and `Checkout.php` inherit from a class that has it turned off, but it's introducing a lot of class stuff just to turn off or on a vendor script.

I've also considered whether we should submit a PR upstream to dependency extraction webpack plugin, my idea is that it would be useful to specify additional dependencies when generating the php asset file.

**Comment by samueljseay** [comment]
CI failure is unrelated

**Comment by nefeline** [comment]
Hey @samueljseay 👋 ! Quick question: currently we have the following as a test instruction for this PR:

> NOTE: Currently Jetpack boost concat JS feature breaks these changes. we need to fix or ask for upstream fix.

Is it correct? Aren't additional tests required? Thanks in advance.

**Comment by samueljseay** [comment]
Apologies @nefeline as we discussed I didn't mean to mark this as ready for review

File: plugins/woocommerce-blocks/bin/add-split-chunk-dependencies.js
Comment (samueljseay) [inline plugins/woocommerce-blocks/bin/add-split-chunk-dependencies.js:13]: This plugin will look through all generated assets with an asset.php file and add split chunks as dependencies to it. The name of the handle will be the configured name in Webpack. in PHP we will still need to register the vendor bundles but we won't have to enqueue them selectively for entry points that don't need them.

The reason I added this was to avoid having to selectively enqueue vendor scripts based on whether a block was a non cart and checkout block or not.

File: plugins/woocommerce-blocks/bin/webpack-entries.js
Comment (samueljseay) [inline plugins/woocommerce-blocks/bin/webpack-entries.js:251]: We still need to generate cart and checkout entry points for the editor side bundle and the styling bundle as normal so we add them back into the entry list here.

File: plugins/woocommerce-blocks/bin/webpack-entries.js
Comment (samueljseay) [inline plugins/woocommerce-blocks/bin/webpack-entries.js:281]: These are the entry points for the new cart and checkout bundle

File: plugins/woocommerce/src/Blocks/AssetsController.php
Comment (samueljseay) [inline plugins/woocommerce/src/Blocks/AssetsController.php:67]: Note that because we suffix files in frontend bundles with `frontend` the name of the asset in Webpack's terms is `wc-blocks-frontend-vendors`, and thats what we append to the asset.php file, but the actual filename is suffixed with `frontend`. I did not want to change anything drastic about how we name or generate files because of i18n and caching, but I have renamed the vendor handle in Webpack so that the file is not named `wc-blocks-vendors-frontend-frontend` which seemed... worse 😅 

I could not call it `wc-blocks-vendors` because that handle is already taken by another bundle that does **not** suffix the filename. If we had some foresight in the past maybe we would not have suffixed at all, but named the vendors more uniquely.

File: plugins/woocommerce/src/Blocks/AssetsController.php
Comment (samueljseay) [inline plugins/woocommerce/src/Blocks/AssetsController.php:72]: We don't need to manually add dependencies any more because we add them to the asset.php files.

**Comment by samueljseay** [comment]
@opr @senadir This is ready for review 🎉

**Comment by samueljseay** [comment]
Just running in to some issues with prod builds, I'm surprised they were not caught by e2e, it may be that e2e is not running against a prod build of JS.

File: plugins/woocommerce/src/Blocks/BlockTypes/AbstractBlock.php
Comment (samueljseay) [inline plugins/woocommerce/src/Blocks/BlockTypes/AbstractBlock.php]: Still working on confirming this, but it looks like we were not enqueueing the dependencies of entry points before (this code just loads an empty array and I don't know why this code is run at all here, maybe some refactoring happened at some point). I've changed this to actually merge in the list of dependencies in the asset.php file that is generated per block.

Comment (samueljseay) [reply plugins/woocommerce/src/Blocks/BlockTypes/AbstractBlock.php]: Ok I think the reason I saw this was because we cache asset files lol. I was testing in a JN site and installing a new version of same plugin but not clearing transient cache. 

Still curious why this code is like this but I'll look at reverting this.

**Comment by opr** [comment]
@samueljseay I really appreciate the work here. This PR covers a lot and I feel like a pairing session to review it would be helpful. I also don't have a deep knowledge of writing Webpack plugins, so you may want to get additional reviews from folks who do.

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (senadir) [inline plugins/woocommerce-blocks/bin/webpack-configs.js:1008]: This solved it mainly for local dev, actual releases don't have this problem because translation is regenerated on each release.

File: plugins/woocommerce-blocks/bin/add-split-chunk-dependencies.js
Comment (senadir) [reply plugins/woocommerce-blocks/bin/add-split-chunk-dependencies.js:13]: I'd love an actual example of what we're solving here, as the text feels a bit too abstract for me.

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (senadir) [inline plugins/woocommerce-blocks/bin/webpack-configs.js:1084]: What happens when we introduce a new folder to the mix here, which is a common process? would things break, or would bundle size increase slightly in a non-optimal way?

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (senadir) [inline plugins/woocommerce-blocks/bin/webpack-configs.js:1109]: I assume this was always there?

File: plugins/woocommerce/src/Blocks/AssetsController.php
Comment (senadir) [reply plugins/woocommerce/src/Blocks/AssetsController.php:67]: It's fine to change names, translation will only be impacted locally and those scripts don't have much strings in them.

**Review by senadir — APPROVED** [review]
I'm loving the save in size! I left some comments to fully understand what's happening and to say that naming changes are fine here.

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-configs.js:1084]: @senadir nothing would break, if that was the case I would feel uncomfortable to manually declare a complex regex.

>  would bundle size increase slightly in a non-optimal way?

Yes it's possible there would be room to improve the bundle size again assuming it was a dependency that's repeated over many entry points in the bundle. I am hoping with the new action that reports JS bundle sizes that we'd spot this kind of thing at the PR stage. I also think its likely that this won't be a big issue because this configuration covers exports that tend to be imported across different entry points whereas what's not covered by this regex tends to be stuff that's specific to only one or 2 entrypoints and wouldn't save KB to be split out here anyway.

File: plugins/woocommerce-blocks/bin/add-split-chunk-dependencies.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/add-split-chunk-dependencies.js:13]: @senadir thanks thats good feedback.

I wonder if I can simplify the wording. I was trying to give an explicit description of the problem here, but maybe I need a concrete example of some kind.

The WP Dependency Extraction plugin does not add split chunks as dependencies to the `.asset.php` files that list dependencies. You'll see that we currently manually enqueue those dependencies in PHP. 

For every generated `.asset.php` file in the whole bundle, this plugin prefixes the list of dependencies with the handles of split chunks that were generated. 

e.g. if your bundle has a vendors script called `foo-vendors` then for every entry-point that has a `.asset.php` file generated by DEWP we will edit that file to include `foo-vendors` as a listed dependency.

I don't want to get too into the weeds here,  but there is a downside to this too, it is possible an entrypoint could have no dependency on a script that is included in `foo-vendors`. **but**, this is not worse than the current state of things where we always enqueue the vendor scripts. 

I would like to be able to use Webpack's chunk graph to determine if an entrypoint has a dependency on a chunk, but after trying a few techniques I found this was very difficult to do, and I'd need to understand more about the chunk graph API to solve that.

Before I added this code, I had to find a place to manually enqueue vendor chunks that made sense, but with this new bundle we now have blocks that are part of the c&c bundle and blocks that are not. Adding the dependency to the `asset.php` file means we don't have to worry about determining if a block is c&c or not and we only enqueue the vendor scripts we need.

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-configs.js:1008]: @senadir This bundle is basically copy/pasted from the front-end bundle including this comment, do you think I should remove this comment or adjust it?

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-configs.js:1109]: > I assume this was always there?

@senadir which part? I did copy paste this from the front-end bundle because the configurations make sense to be exactly the same. The duplication is not ideal, and I think we could do some refactoring of the shared webpack plugins in future.

**Comment by samueljseay** [comment]
@Aljullu Would you mind being an additional reviewer for this one? I just want someone else who has worked on the Webpack stuff to take a look at it.

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (Aljullu) [reply plugins/woocommerce-blocks/bin/webpack-configs.js:1109]: > The duplication is not ideal, and I think we could do some refactoring of the shared webpack plugins in future.

+1 to this. While probably not a blocker of this PR, I would really like that everything that needs to be the same between the frontend bundle and the Cart & Checkout bundle is stored in the same place. Otherwise we risk introducing changes to one bundle and not the other.

File: plugins/woocommerce-blocks/bin/webpack-entries.js
Comment (Aljullu) [inline plugins/woocommerce-blocks/bin/webpack-entries.js]: Have you evaluated if it would be worth including the Mini-Cart Contents block in this bundle? It shares many components and dependencies with the Cart block.

File: plugins/woocommerce-blocks/bin/webpack-entries.js
Comment (Aljullu) [inline plugins/woocommerce-blocks/bin/webpack-entries.js:285]: Related to my comment above, I'm not sure to understand why `mini-cart-component` is added as part of the `cartAndCheckoutFrontend` entry while the Mini Cart and Mini Cart Contents block are not.

**Review by Aljullu — APPROVED** [review]
As usual, thanks for working on this and all the performance improvements these changes are introducing! :clap: 

I'm not an expert on Webpack either, but I took a look at the code and did some smoke testing here and there and couldn't spot anything off. I left some inline comments and questions below, but nothing problematic.

I would appreciate it if @senadir can take a last look before merging, specially to the threads he started.

> Note we will need to add some kind of dev advisory to the release this ends up in, to let people know they need to be on Jetpack Boost 3.4.0+ to ensure they don't run into issues with the "Concatenate JS" feature. We can't workaround that requirement here because we introduce new vendor bundles.

 Does it mean that if the store is using Jetpack Boost < 3.4.0 there will be incompatibilities with the blocks? An alternative option would be to show a notice or something similar asking store owners to update Jetpack Boost. But maybe that's overengineering and we should assume people will understand they need to update all plugins if they encounter issues? :thinking:

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-configs.js:1109]: @Aljullu Yes you're right, I'm quite happy to add a follow up task to tackle refactoring the webpack configs. I'll add an issue for that.

File: plugins/woocommerce-blocks/bin/webpack-entries.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-entries.js:285]: @Aljullu Yes its a good point. I am pretty sure I had a reason at the time because I remember exploring this, but I will have to look again because I can't remember the reason.

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-configs.js:1109]: I've created an issue to deal with that: https://github.com/woocommerce/woocommerce/issues/48968

File: plugins/woocommerce-blocks/bin/add-split-chunk-dependencies.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/add-split-chunk-dependencies.js:13]: I've tried to improve the comment based on what I said in the above comment https://github.com/woocommerce/woocommerce/pull/48010/commits/b1b72e3ef5b358ce887aeba0ca2afb84da0f61d2

File: plugins/woocommerce-blocks/bin/webpack-entries.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-entries.js:285]: I've added mini cart to the bundle and don't see any issues :)

**Comment by samueljseay** [comment]
note that the PHP tests failing is unrelated.

**Comment by samueljseay** [comment]
@Aljullu 

> Does it mean that if the store is using Jetpack Boost < 3.4.0 there will be incompatibilities with the blocks?

Only with the concat JS feature enabled, but yeah, its an issue that boost patched for us last time 

>  An alternative option would be to show a notice or something similar asking store owners to update Jetpack Boost. But maybe that's overengineering and we should assume people will understand they need to update all plugins if they encounter issues? 🤔

I'm not sure, I will ask more broadly about this to get some opinions.

File: plugins/woocommerce-blocks/bin/webpack-entries.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-entries.js]: Done in https://github.com/woocommerce/woocommerce/pull/48010/commits/c30edd911149c1b27692e0b5a2207210f362238e thanks for calling this out.

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (senadir) [reply plugins/woocommerce-blocks/bin/webpack-configs.js:1109]: > I did copy paste this from the front-end bundle because the configurations make sense to be exactly the same.

Yeah that was my question, if it's a new code or existing one moved there.

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (senadir) [reply plugins/woocommerce-blocks/bin/webpack-configs.js:1008]: No it's fine! we can clean up in a follow up if needed.

**Comment by senadir** [comment]
All good from my side!

**Review by senadir — APPROVED** [review]

**Comment by samueljseay** [comment]
@senadir I've added a workaround for old versions of Boost (and updated test instructions) would you mind taking one more look? 🙇🏻 

Or @Aljullu ?

**Review by Aljullu — APPROVED** [review]
I took another look at the code, did some more smoke testing and tested with Jetpack Boost 3.3.1 and everything seems to be working well. LGTM, thanks for all the work on this, @samueljseay!

**Pull request merged**
